apps/docs-snippets/src/guide/transactions/transaction-request.test.ts

```diff
@@ -0,0 +1,191 @@
+    const resource = coin;
+    const resources: Resource[] = [resource];
+
+    // #region transaction-request-3
+    // #import { ScriptTransactionRequest, BaseAssetId };
+
+    // Instantiate the transaction request
+    const transactionRequest = new ScriptTransactionRequest({ script: scriptBytecode });
+
+    // Adding resources (coins or messages)
+    transactionRequest.addResources(resources);
+    transactionRequest.addResource(resource);
+
+    // Adding coin inputs and outputs (including transfer to recipient)
+    transactionRequest.addCoinInput(coin);
+    transactionRequest.addCoinOutput(recipientAddress, 1000, BaseAssetId);
+
+    // Adding message inputs
+    transactionRequest.addMessageInput(message);
+    // #endregion transaction-request-3
+
+    expect(transactionRequest.script).toEqual(arrayify(scriptBytecode));
+    expect(transactionRequest.inputs.length).toEqual(4);
+    expect(transactionRequest.outputs.length).toEqual(2);
+    expect(transactionRequest.witnesses.length).toEqual(1);
+  });
+
+  it('adds a contract to a transaction request', () => {
+    const contractId = Address.fromRandom();
+
+    // #region transaction-request-4
+    // #import { ScriptTransactionRequest };
+
+    // Instantiate the transaction request
+    const transactionRequest = new ScriptTransactionRequest({ script: scriptBytecode });
+
+    // Add the contract input and output using the contract ID
+    transactionRequest.addContractInputAndOutput(contractId);
+    // #endregion transaction-request-4
+
+    expect(transactionRequest.inputs.length).toEqual(1);
+    expect(transactionRequest.outputs.length).toEqual(1);
+  });
+
+  it('adds a predicate to a transaction request', () => {
+    // #region transaction-request-5
+    // #import { ScriptTransactionRequest, Predicate };
+
+    // Instantiate the transaction request
+    const transactionRequest = new ScriptTransactionRequest({ script: scriptBytecode });
+
+    // Instantiate the predicate
+    const predicate = new Predicate({ bytecode: predicateBytecode, abi: predicateAbi, provider });
+
+    // Add the predicate input and resources
+    transactionRequest.addPredicateResource(coin, predicate);
+    // #endregion transaction-request-5
+
+    expect(transactionRequest.inputs.length).toEqual(1);
+    expect(transactionRequest.outputs.length).toEqual(1);
+  });
+
+  it('adds a witness to a transaction request', async () => {
+    const witness = ZeroBytes32;
+
+    // #region transaction-request-6
+    // #import { ScriptTransactionRequest, Account, WalletUnlocked };
+
+    // Instantiate the transaction request
+    const transactionRequest = new ScriptTransactionRequest({ script: scriptBytecode });
+
+    // Add a witness directly
+    transactionRequest.addWitness(witness);
+
+    // Add a witness using an account
+    const account: Account = WalletUnlocked.generate({ provider });
+    await transactionRequest.addAccountWitnesses(account);
+    // #endregion transaction-request-6
```

Comment:
The docs-related text should explicitly say that after the TX is signed at:

```ts
await transactionRequest.addAccountWitnesses(account);
```

Any changes to it will result in the signing not working anymore, because modifying the TX will result in modifying its ID ( only the witnesses property can be changed )

So before signing it, the user needs to add everything that the TX needs, including funds

```ts
    // Instantiate the transaction request
    const transactionRequest = new ScriptTransactionRequest({ script: scriptBytecode });

    transactionRequest.addCoinOutput(recipientAddress, 1000, BaseAssetId);

    const { gasUsed, requiredQuantities, maxFee } = provider.getTransactionCost(
      transactionRequest, [], { signatureCallback: (tx) => tx.addAccountWitnesses(account)}
     );

    transactionRequest.gasLimit = gasUsed;

     // Adding resources (coins or messages)
    await senderWallet.fund(transactionRequest, requiredQuantities, maxFee);

    // Add a witness directly
    transactionRequest.addWitness(witness);

    // Add a witness using an account
    const account: Account = WalletUnlocked.generate({ provider });
    await transactionRequest.addAccountWitnesses(account);
    // #endregion transaction-request-6
```

Since this case is more tricky, it is worth adding a complete test: using a script that requires a signature to make the TX succeed. By doing so, if any changes related to this flow happens, we will remember to update this doc section.

Reply:
So I have added a note [here](https://github.com/FuelLabs/fuels-ts/pull/1961/commits/f622966217675595eb83db74f79141bc5e8b6e97).

The more complete example around tx signing remains at `apps/docs/src/guide/cookbook/transactions-with-multiple-signers.md`. I think that suffices but I will link that here. Do you think that is enough? The example above is mainly to demonstrate the methods accessible to a tx request.

Reply:
https://github.com/FuelLabs/fuels-ts/pull/1961/commits/ddb2f9a287608a5c5b2d970a912dc462c9088a27